#### Data: Car Evaluation

In [1]:
import pandas as pd
import numpy as np

from ucimlrepo import fetch_ucirepo 
  
# fetch dataset 
car_evaluation = fetch_ucirepo(id=19) 
  
# data (as pandas dataframes) 
X = car_evaluation.data.features 
y = car_evaluation.data.targets 

In [2]:
X.shape, y.shape

((1728, 6), (1728, 1))

In [3]:
# concating the two data extracted from the source
data = pd.concat([X,y], axis = 1)

In [4]:
data.shape

(1728, 7)

In [5]:
data.head(3)

buying  maint doors persons lug_boot safety  class
0  vhigh  vhigh     2       2    small    low  unacc
1  vhigh  vhigh     2       2    small    med  unacc
2  vhigh  vhigh     2       2    small   high  unacc

In [6]:
data.columns

Index(['buying', 'maint', 'doors', 'persons', 'lug_boot', 'safety', 'class'], dtype='object')

In [7]:
data.isnull().sum()

buying      0
maint       0
doors       0
persons     0
lug_boot    0
safety      0
class       0
dtype: int64

In [8]:
data['class'].value_counts()

class
unacc    1210
acc       384
good       69
vgood      65
Name: count, dtype: int64

In [9]:
unseen_data = data.sample(frac=0.15, random_state=42)
df = data.drop(unseen_data.index)

In [10]:
unseen_data.to_csv('./Car_Eva/Car_Eva_unseen.csv', index=False)
df.to_csv('./Car_Eva/CarEvaluation.csv', index=False)

In [11]:
data = pd.read_csv('./Car_Eva/CarEvaluation.csv')

In [12]:
data.head(1)

buying  maint doors persons lug_boot safety  class
0  vhigh  vhigh     2       2    small    low  unacc

In [13]:
from pycaret.classification import ClassificationExperiment
s = ClassificationExperiment()
s.setup(data, target = 'class', session_id = 42)

In [14]:
best = s.compare_models()

In [15]:
s.models()

Name  \
ID                                          
lr                    Logistic Regression   
knn                K Neighbors Classifier   
nb                            Naive Bayes   
dt               Decision Tree Classifier   
svm                   SVM - Linear Kernel   
rbfsvm                SVM - Radial Kernel   
gpc           Gaussian Process Classifier   
mlp                        MLP Classifier   
ridge                    Ridge Classifier   
rf               Random Forest Classifier   
qda       Quadratic Discriminant Analysis   
ada                  Ada Boost Classifier   
gbc          Gradient Boosting Classifier   
lda          Linear Discriminant Analysis   
et                 Extra Trees Classifier   
lightgbm  Light Gradient Boosting Machine   
catboost              CatBoost Classifier   
dummy                    Dummy Classifier   

                                                  Reference  Turbo  
ID                                                                  
lr        sklearn.linear_model._logistic.LogisticRegression   True  
knn       sklearn.neighbors._classification.KNeighborsCl...   True  
nb                           sklearn.naive_bayes.GaussianNB   True  
dt             sklearn.tree._classes.DecisionTreeClassifier   True  
svm       sklearn.linear_model._stochastic_gradient.SGDC...   True  
rbfsvm                             sklearn.svm._classes.SVC  False  
gpc       sklearn.gaussian_process._gpc.GaussianProcessC...  False  
mlp       sklearn.neural_network._multilayer_perceptron....  False  
ridge           sklearn.linear_model._ridge.RidgeClassifier   True  
rf          sklearn.ensemble._forest.RandomForestClassifier   True  
qda       sklearn.discriminant_analysis.QuadraticDiscrim...   True  
ada       sklearn.ensemble._weight_boosting.AdaBoostClas...   True  
gbc         sklearn.ensemble._gb.GradientBoostingClassifier   True  
lda       sklearn.discriminant_analysis.LinearDiscrimina...   True  
et            sklearn.ensemble._forest.ExtraTreesClassifier   True  
lightgbm                    lightgbm.sklearn.LGBMClassifier   True  
catboost                   catboost.core.CatBoostClassifier   True  
dummy                         sklearn.dummy.DummyClassifier   True

In [16]:
lightgbm = s.create_model('lightgbm')

In [17]:
lightgbm

LGBMClassifier(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
               importance_type='split', learning_rate=0.1, max_depth=-1,
               min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
               n_estimators=100, n_jobs=-1, num_leaves=31, objective=None,
               random_state=42, reg_alpha=0.0, reg_lambda=0.0, subsample=1.0,
               subsample_for_bin=200000, subsample_freq=0)

In [18]:
tune_lightgbm = s.tune_model(lightgbm)

Fitting 10 folds for each of 10 candidates, totalling 100 fits
Original model was better than the tuned model, hence it will be returned. NOTE: The display metrics are for the tuned model (not the original one).


In [19]:
final_lightgbm = s.finalize_model(tune_lightgbm)

In [20]:
final_lightgbm

Pipeline(memory=Memory(location=None),
         steps=[('label_encoding',
                 TransformerWrapperWithInverse(exclude=None, include=None,
                                               transformer=LabelEncoder())),
                ('numerical_imputer',
                 TransformerWrapper(exclude=None, include=[],
                                    transformer=SimpleImputer(add_indicator=False,
                                                              copy=True,
                                                              fill_value=None,
                                                              keep_empty_features=False,
                                                              missing_values=nan,
                                                              strategy='mean'))),
                ('...
                 LGBMClassifier(boosting_type='gbdt', class_weight=None,
                                colsample_bytree=1.0, importance_type='split',
                                learning_rate=0.1, max_depth=-1,
                                min_child_samples=20, min_child_weight=0.001,
                                min_split_gain=0.0, n_estimators=100, n_jobs=-1,
                                num_leaves=31, objective=None, random_state=42,
                                reg_alpha=0.0, reg_lambda=0.0, subsample=1.0,
                                subsample_for_bin=200000, subsample_freq=0))],
         verbose=False)

In [21]:
s.evaluate_model(final_lightgbm)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Pipeline Plot', 'pipelin…

In [22]:
data = s.predict_model(final_lightgbm)

In [23]:
data[data['class'] !=data['prediction_label']]

Empty DataFrame
Columns: [buying, maint, doors, persons, lug_boot, safety, class, prediction_label, prediction_score]
Index: []

In [24]:
data = pd.read_csv('./Car_Eva/CarEvaluation.csv')

In [25]:
data.head(1)

buying  maint doors persons lug_boot safety  class
0  vhigh  vhigh     2       2    small    low  unacc

In [26]:
from pycaret.classification import setup, models, create_model, tune_model, finalize_model, save_model, evaluate_model
cls = setup(data, target='class', session_id = 42)

In [27]:
lightgbm = create_model('lightgbm')
lightgbm

LGBMClassifier(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
               importance_type='split', learning_rate=0.1, max_depth=-1,
               min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
               n_estimators=100, n_jobs=-1, num_leaves=31, objective=None,
               random_state=42, reg_alpha=0.0, reg_lambda=0.0, subsample=1.0,
               subsample_for_bin=200000, subsample_freq=0)

In [28]:
tuned_lightgbm = tune_model(lightgbm)

Fitting 10 folds for each of 10 candidates, totalling 100 fits
Original model was better than the tuned model, hence it will be returned. NOTE: The display metrics are for the tuned model (not the original one).


In [29]:
final_lightgbm = finalize_model(tuned_lightgbm)
final_lightgbm

Pipeline(memory=Memory(location=None),
         steps=[('label_encoding',
                 TransformerWrapperWithInverse(exclude=None, include=None,
                                               transformer=LabelEncoder())),
                ('numerical_imputer',
                 TransformerWrapper(exclude=None, include=[],
                                    transformer=SimpleImputer(add_indicator=False,
                                                              copy=True,
                                                              fill_value=None,
                                                              keep_empty_features=False,
                                                              missing_values=nan,
                                                              strategy='mean'))),
                ('...
                 LGBMClassifier(boosting_type='gbdt', class_weight=None,
                                colsample_bytree=1.0, importance_type='split',
                                learning_rate=0.1, max_depth=-1,
                                min_child_samples=20, min_child_weight=0.001,
                                min_split_gain=0.0, n_estimators=100, n_jobs=-1,
                                num_leaves=31, objective=None, random_state=42,
                                reg_alpha=0.0, reg_lambda=0.0, subsample=1.0,
                                subsample_for_bin=200000, subsample_freq=0))],
         verbose=False)

In [30]:
save_model(final_lightgbm, './Car_Eva/Car_Eva_Unseen')

Transformation Pipeline and Model Successfully Saved


(Pipeline(memory=Memory(location=None),
          steps=[('label_encoding',
                  TransformerWrapperWithInverse(exclude=None, include=None,
                                                transformer=LabelEncoder())),
                 ('numerical_imputer',
                  TransformerWrapper(exclude=None, include=[],
                                     transformer=SimpleImputer(add_indicator=False,
                                                               copy=True,
                                                               fill_value=None,
                                                               keep_empty_features=False,
                                                               missing_values=nan,
                                                               strategy='mean'))),
                 ('...
                  LGBMClassifier(boosting_type='gbdt', class_weight=None,
                                 colsample_bytree=1.0, importance_type='split

In [31]:
import joblib

from sklearn.metrics import classification_report
test = pd.read_csv('./Car_Eva/Car_Eva_unseen.csv')

In [32]:
test.shape

(259, 7)

In [33]:
cls = joblib.load('./Car_Eva/Car_Eva_Unseen.pkl')
cls

Pipeline(memory=Memory(location=None),
         steps=[('label_encoding',
                 TransformerWrapperWithInverse(exclude=None, include=None,
                                               transformer=LabelEncoder())),
                ('numerical_imputer',
                 TransformerWrapper(exclude=None, include=[],
                                    transformer=SimpleImputer(add_indicator=False,
                                                              copy=True,
                                                              fill_value=None,
                                                              keep_empty_features=False,
                                                              missing_values=nan,
                                                              strategy='mean'))),
                ('...
                 LGBMClassifier(boosting_type='gbdt', class_weight=None,
                                colsample_bytree=1.0, importance_type='split',
                                learning_rate=0.1, max_depth=-1,
                                min_child_samples=20, min_child_weight=0.001,
                                min_split_gain=0.0, n_estimators=100, n_jobs=-1,
                                num_leaves=31, objective=None, random_state=42,
                                reg_alpha=0.0, reg_lambda=0.0, subsample=1.0,
                                subsample_for_bin=200000, subsample_freq=0))],
         verbose=False)

In [34]:
unseen_X = test.drop('class', axis=1)
y = test['class']

In [35]:
y_pred = cls.predict(unseen_X)
y_pred

0      unacc
1      unacc
2      unacc
3        acc
4      unacc
       ...  
254    unacc
255    unacc
256    unacc
257    unacc
258      acc
Name: class, Length: 259, dtype: object

In [36]:
print(classification_report(y, y_pred))

              precision    recall  f1-score   support

         acc       1.00      0.93      0.96        59
        good       0.83      1.00      0.91        10
       unacc       0.99      1.00      0.99       179
       vgood       1.00      1.00      1.00        11

    accuracy                           0.98       259
   macro avg       0.96      0.98      0.97       259
weighted avg       0.99      0.98      0.98       259



This classification result shows how well this model performed in predicting different classes in Car Evaluation dataset. Here's the explanation of each metric:

Precision: It measures the accuracy of the positive predictions made by the modeli.ele, for class 'acc', the precision is 1.00, which means that all instances predicted as 'acc' were correct.

Recall: It measures the ability of the model to correctly identify all positive instanci.emple, for class 'good', the recall is 1.00, which means that all actual instances of 'good' were correctly predicted by the model.

F1-score: It is the harmonic mean of precision and recall, providing a single score that balances both metrics. It's usefulween you want to consider both precision and recall together.

Support: It indicates the number of actual instances of each cCar Evaluations in the dataset.

Accuracy: It measures the overall correctness of the model across allthereforesses. Here, the accuracy is 0.98, which means that the model correctly predicted 98% of all instances in the dataset.

The "macro avg" and "weighted avg" are the averages of precision, recall, and F1-score across all classes. "Macro avg" gives equal weight to each class, while "weighted avg" takes into account the class imbalance by weighting each class's score by its support.

Overall, this classification result suggests that the model has excellent performance, with high precision, recall, and F1-score for most classes, and a high overall accuracy of 98%.